core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java

```diff
@@ -849,4 +856,118 @@ protected Query createFieldQuery(Analyzer analyzer, BooleanClause.Occur operator
             throw new RuntimeException("Error analyzing query text", e);
         }
     }
+
+    /**
+     * See {@link MapperQueryParser#analyzeGraphPhraseWithLimit}
+     */
+    @Override
+    protected SpanQuery analyzeGraphPhrase(TokenStream source, String field, int phraseSlop) throws IOException {
+        if (shouldApplyGraphPhraseLimit()) {
+            return analyzeGraphPhraseWithLimit(source, field, phraseSlop, this::createSpanQuery);
+        } else {
+            return super.analyzeGraphPhrase(source, field, phraseSlop);
+        }
+    }
+
+    /** A BiFuntion that can throw an IOException */
+    @FunctionalInterface
+    public interface CheckedBiFunction<T, U, R> {
+
+        /**
+         * Applies this function to the given arguments.
+         *
+         * @param t the first function argument
+         * @param u the second function argument
+         * @return the function result
+         */
+        R apply(T t, U u) throws IOException;
+    }
+
+    /**
+     * Checks the value of the JVM option <code>es.query.write.apply_graph_phrase_limit</code> to determine
+     * if the analysis of graph phrase should be limited to {@link BooleanQuery#getMaxClauseCount()}.
+     * The JVM option can only be set to <code>true</code> (false is the default value), any other value
+     * will throw an {@link IllegalArgumentException}.
+     */
+    public static boolean shouldApplyGraphPhraseLimit() {
+        String value = System.getProperty("es.query.apply_graph_phrase_limit");
+        if (value == null) {
+            return false;
+        } else if (Booleans.parseBoolean(value, false) == false) {
+            throw new IllegalArgumentException("[" + value + "] is not a valid value for the JVM option:" +
+                "[es.query.apply_graph_phrase_limit]. Set it to [true] to activate the limit.");
```

Comment:
Is there a way we can check this property on startup as well/instead? Maybe we could check it in SearchModule or MapperService?

The concern I see here is that restarts can be a pain for users and here they will only get feedback on the setting when they try to use a span query and then they can't use the query until they get the admin to restart the cluster

Reply:
I also wonder if the user does not set the limit and we expand beyond the max clause count if we should emit a deprecation warning telling the user that this limit will be enforced in future versions?

Reply:
Good idea, I pushed https://github.com/elastic/elasticsearch/pull/34061/commits/5bd6406b49aa44d3b6f2d3e8fe6936232ff41042 to add the deprecation warning if the system property is not set. It also adds a validation of the system property (only `true` is valid) in `SearchModule`